### ISSUE TITLE
Add filename and line number support to tags, ala __FILE__ and __LINE__ macros

### ISSUE THREAD

**Issue by iwolf81** [doxygen/doxygen#7946]
Hello,

I'd like to tag a line of code such that the HTML output identifies the filename and line number in a manner similar to what the C/C++ preprocessor does with the \_\_FILE\_\_ and \_\_LINE\_\_ macros. 

For example, given the following code segment:

foo.c:
line 965:  bar();  /// @test{file,lineno} Suite A : TC 1 : BP 2

I'd like to see Doxygen generate the following for the Test List output:

Global bar (void)
foo.c:864 Suite A : TC 1: BP 2

By identifying the filename and line number in this manner, I could create a tool that automatically updates the filenames and line number for breakpoints in debugger scripts when the source files are modified. The same capability could be used to create and automatically update a Software Design Document to Source Code traceability matrix, as suggested by @ls6777 in [Doxygen Create Link to specific line of code](https://stackoverflow.com/questions/59327898/doxygen-create-link-to-specific-line-of-code).

Thank you,
Ira

**Comment by thom-sch** [comment]
I'd also like to ask for an enhancement of Doxygen like above.

I'm  working on documenting Delphi source codes using a set of ALIASES.

One of the main goals was to have included the Delphi source code snippets
as well as the C++ *.hpp snippets (generated by Delphi for C++ builder).
So most of the ALIASES have a parameter which refers to the
currently parsed source file, as an (shortened) example ALIAS DClass:

  \DClass{2}="\class \1 ^^ ... \dontinclude \2.pas ^^ \snippet ...source code snippet... "
  with
  - \1 = class name to document
  - \2 = current source file's name (without extension .pas)

There are some more ALIASES like this to document other entities (types, functions etc.),
and each of them gets the source file's name as parameter.

Things would be much easier if there would be tags as mentioned above like:
  - __FILENAME__: file name only, with extension
  - __FILE__: file name only, without extension
  - __FILEDIR__: file's directory
  - __FILEPATH__: complete file path

In this case parameter \2 of ALIAS \DClass (and all other ALIASES) could be
omitted resulting in better readability and avoiding to repeat the source
file's name in every other ALIAS, in this case:

  \DClass{1}="\class \1 ^^ ... \dontinclude __FILE__.pas ^^ \snippet ...source code snippet... "

Regards from Meerbusch
Thomas Schneider

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9461

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by thom-sch** [comment]
Hello,
thanks for adding this feature; trying to use is i've found two small issues:

1. The documentation lists `file` as one of the options but I've figured out that this should be `name` instead. Omitting the option or using invalid options obviously default to `filename`.
2. `\fileinfo` (and probably `lineinfo` - I've not tested) generates an additional space at the end - which prevents concatenating of further info, f.e. referring to `module.pas` from inside `module.cpp` using `\fileinfo{name}.pas` since this was resolved to `module .pas` (note the additional space after module!).

Would be great if this additional space would disappear ;)
Should I open a new issue for this?

Regards from Meerbusch
Thomas

**Comment by albert-github** [comment]
1. The documentation of the option `file` states:
    > with file the name of the file without extension

    This is not correct as in the code I see the line
    ```
    { "name",      [](yyscan_t s,FileInfo &fi) { addOutput(s,fi.baseName());      } },
    ```
    the documentation should be corrected.
    Regarding the default, this should also be added to the documentation.
    See #9547 
2. The problem with the space is indeed also present with the `\lineinfo`  although I cannot directly think of a use case there.
    Best is to create a small test case for this problem.

**Comment by thom-sch** [comment]
Thanks for quick response.

1. I don't agree. The documentaion says: "The option can be `file`, `extension`, `filename`, `directory` or, `full`,..." but in fact the option `file` produces the same output as the option `filename`; using option `name` instead, the `fileinfo`correctly gives the filename without extension.
see attached zip-file

2. The attached zip-file also contains an use case which fails due to the additional space output by `\fileinfo`.

[Test19_fileinfo.zip](https://github.com/doxygen/doxygen/files/9443955/Test19_fileinfo.zip)

Regards from Meerbusch
Thomas

**Comment by albert-github** [comment]
The documentation was incorrect and has been corrected in the mentioned proposed pull request #9547 as this was not correct.
Running the given example (Test19_fileinfo) gives:
```
.../Modul1.md:13: warning: Unknown option specified with \fileinfo: 'file'
.../Modul1.md:14: warning: Unknown option specified with \fileinfo: 'xxxx'
```
so doxygen warns for the wrong `option`s `file` and `xxxx` and will fall back to the default here i.e. `filename`.

**Comment by albert-github** [comment]
Regarding the documentation, update has been integrated in master on GitHub
